### PR TITLE
Handle site extension concurrent installation and corrupted status file.

### PR DESCRIPTION
**Pull request by shrimpy**

### CODE REVIEW

File: Kudu.Core/SiteExtensions/SiteExtensionInstallationLock.cs [hunk 1]
```diff
@@ -5,6 +5,8 @@ namespace Kudu.Core.SiteExtensions
 {
     public class SiteExtensionInstallationLock : LockFile
     {
+        private const string LockNameSuffix = "install.lock";
```
Comment (amitapl) [inline Kudu.Core/SiteExtensions/SiteExtensionInstallationLock.cs]: isn't that the actual lock name not just the suffix?

File: Kudu.Core/SiteExtensions/SiteExtensionBatchUpdateStatusLock.cs [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+        public const string LockNameSuffix = "batch.lock";
+
+        private SiteExtensionBatchUpdateStatusLock(string path)
+            : base(path)
+        { }
+
+        public static SiteExtensionBatchUpdateStatusLock CreateLock(string rootPath)
+        {
+            string lockFilePath = Path.Combine(rootPath, LockNameSuffix);
+            FileSystemHelpers.CreateDirectory(Path.GetDirectoryName(lockFilePath));
```
Comment (amitapl) [inline Kudu.Core/SiteExtensions/SiteExtensionBatchUpdateStatusLock.cs]: ensure directory?

**Comment by amitapl** [comment]
looks good

File: Kudu.Core/SiteExtensions/SiteExtensionStatus.cs [hunk 3]
```diff
@@ -101,6 +100,26 @@ public bool IsTerminalStatus()
                 || string.Equals(Constants.SiteExtensionProvisioningStateCanceled, ProvisioningState, StringComparison.OrdinalIgnoreCase);
         }
 
+        private string SafeRead(string key)
+        {
+            try
+            {
+                return _jsonSettings.GetValue(key);
+            }
+            catch
```
Comment (suwatch) [inline Kudu.Core/SiteExtensions/SiteExtensionStatus.cs]: Catch exception and trace Unexpected exception.

File: Kudu.Core/SiteExtensions/SiteExtensionStatus.cs [hunk 4]
```diff
@@ -101,6 +100,26 @@ public bool IsTerminalStatus()
...
+            {
+                // if setting file happen to be invalid, e.g w3wp.exe was kill while writting
+                // treat it as failed, and suggest user to re-install or un-install
+                _jsonSettings.SetValues(new KeyValuePair<string, JToken>[] {
```
Comment (suwatch) [inline Kudu.Core/SiteExtensions/SiteExtensionStatus.cs]: SetValues also perform read.  You should do the delete the setting files before doing set value.

File: Kudu.Core/SiteExtensions/SiteExtensionInstallationLock.cs (hunk 1 above)
Comment (shrimpy) [inline Kudu.Core/SiteExtensions/SiteExtensionInstallationLock.cs]: correct! each extension will has it own folder. if two request try to install the exact same extension at the same time. we let one in, and reject (409 conflict) the other one.

File: Kudu.Core/SiteExtensions/SiteExtensionBatchUpdateStatusLock.cs (hunk 2 above)
Comment (shrimpy) [inline Kudu.Core/SiteExtensions/SiteExtensionBatchUpdateStatusLock.cs]: ah.. sure.. thanks for noticing that.

File: Kudu.Core/SiteExtensions/SiteExtensionStatus.cs (hunk 3 above)
Comment (shrimpy) [inline Kudu.Core/SiteExtensions/SiteExtensionStatus.cs]: right. this is very important. thank you!

File: Kudu.Core/SiteExtensions/SiteExtensionStatus.cs (hunk 4 above)
Comment (shrimpy) [inline Kudu.Core/SiteExtensions/SiteExtensionStatus.cs]: right. this is very important. thank you!

Comment (shrimpy) [inline Kudu.Core/SiteExtensions/SiteExtensionStatus.cs]: for write, if failed, i will clear all the content, and write the new key/value into json file.

**Pull request merged**